miio/gateway/devices/subdevice.py

```diff
@@ -260,3 +266,67 @@ def get_firmware_version(self) -> Optional[int]:
                 ex,
             )
         return self._fw_ver
+
+    def Register_callback(self, id, callback):
+        """Register a external callback function for updates of this subdevice."""
+        if id in self._registered_callbacks:
+            _LOGGER.error(
+                "A callback with id '%s' was already registed, overwriting previous callback",
+                id,
+            )
+        self._registered_callbacks[id] = callback
+
+    def Remove_callback(self, id):
+        """Remove a external callback using its id."""
+        self._registered_callbacks.pop(id)
+
+    def push_callback(self, action, params):
+        """Push callback received from the push server."""
+        if action not in self.push_events:
+            _LOGGER.error(
+                "Received unregistered action '%s' callback for sid '%s' model '%s'",
+                action,
+                self.sid,
+                self.model,
+            )
+
+        prop = self.push_events[action].get("property")
+        value = self.push_events[action].get("value")
+        if prop is not None and value is not None:
+            self._props[prop] = value
+
+        for callback in self._registered_callbacks.values():
+            callback(action, params)
+
+    def subscribe_events(self):
+        """subscribe to all subdevice events using the push server."""
+        if self._gw._push_server is None:
+            _LOGGER.error("Can not install push callback withouth a push_server")
+            return False
+
+        result = True
+        for action in self.push_events:
+            event_info = EventInfo(
+                action=action,
+                extra=self.push_events[action]["extra"],
+                source_sid=self.sid,
+                source_model=self.zigbee_model,
+                event=self.push_events[action].get("event", None),
+                command_extra=self.push_events[action].get("command_extra", ""),
+                trigger_value=self.push_events[action].get("trigger_value"),
+            )
+
+            event_id = self._gw._push_server.subscribe_event(self._gw, event_info)
+            if event_id is None:
+                result = False
+                continue
```

Comment:
This method will return `False`, even if some of the subscriptions succeeded. Is that expected behavior?

Reply:
Yes, all subscriptions schould succeed otherwise the method schould return False indicating something is wrong.